### PR TITLE
feat: filter variables on cohort variable mapping

### LINKED ISSUE #3193: Filter on cohorts in the SSR catalogue variable view
**Is your feature request related to a problem? Please describe.**
As researcher or datamanager I want to be able to filter on cohorts of interest, so I only see variables that have information for one of the filtered cohorts available.

**Describe the solution you'd like**
Filter for cohorts (within the selected network). When one or more cohorts are selected, only variables that have info for one of the selected cohorts are shown and the harmonization matrix is filtered for these selected cohorts as well.

### PR DESCRIPTION
**Pull request by connoratrug**
Add the ability to filter (mapped) variables by cohort

Closes #3193

- Split list filter into list and filter part
- Move list to component lib, and add demo page
- Add cohort filter to variables page default filter config
- Add option to pass custom filterCondition build function to Filter type
- If cohort filter is active, filter the list by including only variables that either have a direct mapping to the cohort or have a repeat that has a mapping to the cohort.

how to test:
- go to the filters page
- select the cohorts filter -> cohorts should be loaded and shown
- select one or more cohorts 
- variables should now be filters to only include those that have mapping ( source) to the selected cohorts 

todo:
- [x] added tests
- [ ] updated testplan to include a test for this fix, including ref to bug using # notation

### CODE REVIEW

**Review by mswertz — APPROVED** [review]
functional approve. Maybe a second programmer look at the code? Otherwise okay to merge (I scanned through it).

**Comment by connoratrug** [comment]
these seems to be some confusion on what is being shown on the variables list page, https://data-catalogue.molgeniscloud.org/catalogue/ssr-catalogue/all/variables and https://data-catalogue.molgeniscloud.org/catalogue/ssr-catalogue/LongITools/variables.

The title seems to suggest the list includes all variables ( in case of subcatalogue; linked to subcatalogue ). However in https://github.com/molgenis/molgenis-emx2/issues/3193 there is the request to be able to filter the variable list on cohorts that are liked to the variable via a mapping ( see this pr for implementation details). 

This seems confusing/inconsistent, so the question becomes;

- What problem is the screen trying to solve, do we want to show all available variables or all mapped variabels ( and why )?
    - if we only want to show mapped variabels, is this made clear though the current (sub)title (_A complete overview of available variables._)  ( then we would need to change the list query to only show the variables that are used in a mapping)
- Should we provide the option to view either the mapped variabels or all the variables ( are these both relevant in the context of the user story ) 
- If we do want to show all variables ( not only mapped ), should we then change the cohort filter query ? , or make it clear via the filter title and description that applying a cohort filter also now includes a mapping filter ( then we should also change the list filter when a cohort filter is active )

**Comment by marijevdgeest** [comment]
I get your confusion (I think this is really a case where I'm so used to this set up that I don't see this logic anymore). 
I think we need to either decide where we want to show source variables: 
- On the cohort page like we discussed before? (Then it's hard to browse but it's also a feature to have that info in the cohort's context) -> variable explorer needs different titel or explanatory text
- In the variable explorer -> possibly some changes to set up/filters needed.

@EleanorHyde-UMCG @timcadman any thoughts?

**Review by mswertz — COMMENTED** [review]
I get an list of variables with no mappings in this case (might it be that there 'no match' statuses in here?)

image

**Comment by marijevdgeest** [comment]
> I get an list of variables with no mappings in this case (might it be that there 'no match' statuses in here?)
> 
> image

Indeed no data, I see the matrix also disappeared including cohort name

**Comment by connoratrug** [comment]
i'll have a look , please be exact with your 'how to reproduce' ( add steps, expected and actual). need to figure out if its a code issue and or a data issue

**Comment by connoratrug** [comment]
@mswertz @marijevdgeest , i can not reproduce your findings, please add exact steps on how to reproduce

**Review by BrendaHijmans — DISMISSED** [review]
De cohort filter doet nu wat hij moet doen: filteren op de variabelen die vanuit een geselecteerd cohort zijn gemapt. Als de repeats van een variabele zijn gemapt, wordt de 'root' variabele ook getoond, dat klopt ook. Er is echter een inconsistentie nu met de matrix als je doorklikt. In het cohort filter worden namelijk de 'na' mappings meegenomen, maar in de overview per variabele niet. Hieronder in stappen met plaatjes uitgelegd.

1. Ga in de preview naar https://preview-emx2-pr-3656.dev.molgenis.org/catalogue-demo/ssr-catalogue/testNetwork1/variables
2. selecteer testCohort 2, zie dat er twee variabelen zijn gemapt voor dit cohort 
image

3. klik op testVariableRepeats_
4. Zie dat testCohort2 niet tussen de harmonisation overview staat voor deze variabele en zijn repeats 
image
5. Dit komt doordat mappings met 'match' = 'na' wel in de cohort filter worden meegenomen, maar niet in de harmonisation overview. Zie hier de data in Variable mappings tabel: 
image

**Comment by connoratrug** [comment]
> De cohort filter doet nu wat hij moet doen: filteren op de variabelen die vanuit een geselecteerd cohort zijn gemapt. Als de repeats van een variabele zijn gemapt, wordt de 'root' variabele ook getoond, dat klopt ook. Er is echter een inconsistentie nu met de matrix als je doorklikt. In het cohort filter worden namelijk de 'na' mappings meegenomen, maar in de overview per variabele niet. Hieronder in stappen met plaatjes uitgelegd.
> 
> 1. Ga in de preview naar https://preview-emx2-pr-3656.dev.molgenis.org/catalogue-demo/ssr-catalogue/testNetwork1/variables
> 2. selecteer testCohort 2, zie dat er twee variabelen zijn gemapt voor dit cohort
>    image
> 3. klik op testVariableRepeats_
> 4. Zie dat testCohort2 niet tussen de harmonisation overview staat voor deze variabele en zijn repeats
>    image
> 5. Dit komt doordat mappings met 'match' = 'na' wel in de cohort filter worden meegenomen, maar niet in de harmonisation overview. Zie hier de data in Variable mappings tabel:
>    image

This is not related to the current this PR, suggest to submit a issue or feature request

File: apps/nuxt3-ssr/pages/[schema]/ssr-catalogue/[catalogue]/variables/index.vue [hunk 1]
```diff
@@ -64,12 +70,80 @@ const pageFilterTemplate: IFilter[] = [
...
+        filterBuilder: Record<string, Record<string, any>>,
+        conditions: IFilterCondition[]
+      ) => {
+        return {
+          ...filterBuilder,
+          ...{
+            _or: [
+              {
+                mappings: {
+                  source: { equals: conditions.map((c) => ({ id: c.name })) },
```
Comment (mswertz) [inline apps/nuxt3-ssr/pages/[schema]/ssr-catalogue/[catalogue]/variables/index.vue:93]: I believe this should also filter on match = 'partial' or  match = 'complete' (e.g. this data https://preview-emx2-pr-3656.dev.molgenis.org/catalogue-demo/tables/#/VariableMappings?_filter=match&match=[{%22name%22:%22complete%22},{%22name%22:%22partial%22}])

File: apps/nuxt3-ssr/pages/[schema]/ssr-catalogue/[catalogue]/variables/index.vue [hunk 2]
```diff
@@ -64,12 +70,80 @@ const pageFilterTemplate: IFilter[] = [
...
+                },
+              },
+              {
+                repeats: {
+                  mappings: {
+                    source: { equals: conditions.map((c) => ({ id: c.name })) },
```
Comment (mswertz) [inline apps/nuxt3-ssr/pages/[schema]/ssr-catalogue/[catalogue]/variables/index.vue:99]: idem

**Review by mswertz — CHANGES_REQUESTED** [review]

**Comment by EleanorHyde-UMCG** [comment]
If I only select on testcohort2, I should see no variables at all. Now I see testVarNoRepeats and testVarRepeats_.  

The number of cohorts, however, is correct (see discussion on #cohorts_federation Slack - this may solve that problem)

File: apps/nuxt3-ssr/pages/[schema]/ssr-catalogue/[catalogue]/variables/index.vue (hunk 1 above)
Comment (mswertz) [reply apps/nuxt3-ssr/pages/[schema]/ssr-catalogue/[catalogue]/variables/index.vue:93]: solution is to add `match: { name: { equals: [ "complete" ,"partial"] }} `

**Comment by mswertz** [comment]
> If I only select on testcohort2, I should see no variables at all. Now I see testVarNoRepeats and testVarRepeats_.
> 
> The number of cohorts, however, is correct (see discussion on #cohorts_federation Slack - this may solve that problem)

We have pushed a fix for this, please check again

**Comment by EleanorHyde-UMCG** [comment]
> > If I only select on testcohort2, I should see no variables at all. Now I see testVarNoRepeats and testVarRepeats_.
> > The number of cohorts, however, is correct (see discussion on #cohorts_federation Slack - this may solve that problem)
> 
> We have pushed a fix for this, please check again

FIltering on testcohort2 does not show no variables at all.

However, new edge case found right now: the number of cohorts is correct EXCEPT when there are no mappings to be shown. Filter on variable name 'cat' and you get testvarCategorical, with the text "1 variable in 4 cohorts". This should be "1 variable in 0 cohorts" because this variable has not been mapped to in any of the 4 cohorts.

**Comment by EleanorHyde-UMCG** [comment]
> > > If I only select on testcohort2, I should see no variables at all. Now I see testVarNoRepeats and testVarRepeats_.
> > > The number of cohorts, however, is correct (see discussion on #cohorts_federation Slack - this may solve that problem)
> > 
> > 
> > We have pushed a fix for this, please check again
> 
> FIltering on testcohort2 now shows no variables at all, correct.
> 
> However, new edge case found right now: the text with the number of cohorts is correct EXCEPT when there are no mappings to be shown. Filter on variable name 'cat' and you get testvarCategorical, with the text "1 variable in 4 cohorts". This should be "1 variable in 0 cohorts" because this variable has not been mapped to in any of the 4 cohorts.

Have just discussed that we will leave off the 'in xx variables' text to avoid confusion.

**Comment by mswertz** [comment]
> Have just discussed that we will leave off the 'in xx variables' text to avoid confusion.

done

**Review by mswertz — APPROVED** [review]

**Comment by mswertz** [comment]
> Tested and it works correctly.
> 
> One thought is that currently the behaviour when ticking all cohorts is the same as when ticking none. Intuitively I would expect if no cohorts are selected, no data is shown. Could an option be that by default/on initial search all available cohorts ticked, and the user can chose to deselect cohorts which are not needed?
> 
> I'm also happy for this to be classed an enhancement and considered later (perhaps if we get similar feedback from users)

I indeed would like to have that as refinement, i.e., the option to only show variables that have mappings by one or more cohorts vs all variables defined in the network.

**Review by timcadman — APPROVED** [review]

**Pull request merged**